[docs] - SDA guide for existing Dagster users [CON-30] (#8188)
dagster-io/dagster: docs/content/guides/dagster/enriching-with-software-defined-assets.mdx

> # Software-Defined Assets for Existing Dagster Users

Reviewer: This doesn't explicitly apply to existing Dagster users, so to me, this title doesn't accurately convey what this guide is actually doing.

Author: Mind elaborating a little bit on what you mean by this?

Reviewer: Sure. This guide, to me, is more than about existing vs new users - it's about the difference between SDA and ops/graphs/etc. This guide lays out when SDA is a better fit over other approaches, which is useful to both new *and* existing users.

Existing users might want to enrich their jobs (as you say in here!) and new users might not be sure what to do and are looking for guidance.

My concern is that the title calling out `existing users` may make new users feel like it's not applicable to them and then they miss out on some really good info.

I'm also not sure that this title will have the guide show up in a good spot in search results. The question for that is what would someone search for if they were trying to do the very thing this guide describes?

Author: Got it - those points make sense to me, but I'm struggling a little bit to think of a different name that I like better. Do you have one in mind?

One of the things that makes it "existing user"-y is that it assumes a lot of knowledge about existing Dagster concepts. It gives some basic definitions of ops and graphs, but it doesn't go out of its way to explain how they work.

Reviewer: I just pushed up my changes and went with `Upgrading to Software-Defined Assets` to just have something, but I'm open to other suggestions.

`Upgrade` implies that something already exists, so perhaps that could satisfy the existing user part of this? TBH I don't know that a piece of documentation *needs* to explain in full all the concepts it touches on to be new user-friendly.  As long as we link out and give context for what the current piece focuses on, I think that's enough, at least for this!

Here's a job where some of the ops - `extract_products` and `get_categories` don't producing an asset of its own - they just produce transient data that downstream ops will use to produce assets:

```python file=/guides/dagster/enriching_with_software_defined_assets/vanilla_graph.py
from pandas import DataFrame

from dagster import job, op, repository

from .mylib import create_db_connection, fetch_products

@op
def extract_products() -> DataFrame:
    return fetch_products()

@op
def get_categories(products: DataFrame) -> DataFrame:
    return DataFrame({"category": products["category"].unique()})

@op
def write_products_table(products: DataFrame) -> None:
    products.to_sql(name="categories", con=create_db_connection())

@op
def write_categories_table(categories: DataFrame) -> None:
    categories.to_sql(name="categories", con=create_db_connection())

@job
def ingest_products_and_categories():
    products = extract_products()
    product_categories = get_categories(products)
    return write_products_table(products), write_categories_table(product_categories)

@repository
def repo():
    return [ingest_products_and_categories]
```

> Here's an equivalent job with software-defined assets. Note that, because not all the ops correspond to assets, it makes use of the `@op` and `@graph` APIs and uses `from_graph` to wrap a graph in a software-defined asset.

Reviewer: ```suggestion
Here's an equivalent job with software-defined assets. Note that because some ops don't correspond to assets, this job uses `@op` and `@graph` APIs and `from_graph` to wrap a graph in a software-defined asset:
```

I wanted to improve the clarity here, so I'm suggesting some grammar changes. Most importantly replacing `it` with `the job` so the reference is clear.

```python file=/guides/dagster/enriching_with_software_defined_assets/sda_graph.py
from pandas import DataFrame

from dagster import AssetSelection, AssetsDefinition, GraphOut, graph, op, repository

from .mylib import create_db_connection, fetch_products

@op
def extract_products() -> DataFrame:
    return fetch_products()

@op
def get_categories(products: DataFrame) -> DataFrame:
    return DataFrame({"category": products["category"].unique()})

@op
def write_products_table(products: DataFrame) -> None:
    products.to_sql(name="categories", con=create_db_connection())

@op
def write_categories_table(categories: DataFrame) -> None:
    categories.to_sql(name="categories", con=create_db_connection())

@graph(out={"products": GraphOut(), "categories": GraphOut()})
def ingest_graph():
    products = extract_products()
    product_categories = get_categories(products)
    return write_products_table(products), write_categories_table(product_categories)

two_tables = AssetsDefinition.from_graph(ingest_graph)

@repository
def repo():
    return [two_tables, AssetSelection.all().to_job("products_and_categories")]
```

> ## Future work: jobs that run ops after assets

Reviewer: Including this as-is feels out of place to me. I'm not sure what to do with it at the moment, but I'll give it some thought.

In the future, you'll be able to define jobs that materialize software-defined assets and then run arbitrary ops afterwards. If this is something you think you might use, we'd love to hear from you in Slack or in a Github discussion.

> ## Concept mapping

Reviewer: This section is great, but it feels odd to have it at the end here. To me, that makes it feel sorta thrown in, when it actually contains some really useful info.

Perhaps we should move it up and re-name it - something like `What are the differences between ops, graphs, and software-defined assets?`

Author: The title "Concept mapping" isn't a great one.  What I'm trying to get across here isn't about the differences though. The way I imagined it fits into the overall narrative is that the earlier parts of the blog post give the broad strokes of how to migrate in common cases.  This is covering the long tail.  The aim is to answer questions like "If I start using software-defined assets, does that mean I no longer get to use config?"

Reviewer: Gotcha. I'll think on this a bit.

---
title: Upgrading To Software-Defined Assets | Dagster
description: Understand when, why, and how to use software-defined assets (SDAs) in Dagster, as well as how SDAs work with other core Dagster concepts.
---

# Upgrading To Software-Defined Assets

> If you want to understand when, why, and how to use software-defined assets in Dagster, this guide is for you. We'll also demonstrate what some common Dagster jobs look like before and after using software-defined assets.

Author: Based on the first sentence, I think it's a little difficult to know what to expect for how the content in here would be different than the content in - say - the SDA concepts page. I think it's worth trying to target it a little bit more specifically. I can brainstorm options with you if it would be helpful.

Reviewer: > If you're familiar with ops and graphs and want to understand when, why, and how to use software-defined assets in Dagster, this guide is for you. We'll also demonstrate what some common Dagster jobs look like before and after using software-defined assets.

Before we jump in, here's a quick refresher:

- An **asset** is a persistent object in storage, such as a table, machine learning (ML) model, or file.
- An [**op**](/concepts/ops-jobs-graphs/ops) is the core unit of computation in Dagster. For example, an op might accept tabular data as input and produce transformed tabular data as output.
- A [**graph**](/concepts/ops-jobs-graphs/jobs-graphs) is a directed acyclic graph of ops or other graphs, which execute in order and pass data to each other.
- A [**software-defined asset**](/concepts/assets/software-defined-assets) is a declaration of an asset that should exist and a description of how to compute it: the op or graph that needs to run and the upstream assets that it should run on.

> **Software-defined assets aren't a replacement for Dagster's core computational concepts**. Think of them as a top layer that links ops, graphs, and jobs to the long-lived objects they interact with.

Reviewer 2: I think it would be helpful here to reiterate that ops are the core unit of computation that occurs within an asset.

### Lineage

As software-defined assets know what other assets they depend on, an asset's lineage can be [viewed easily in Dagit](/concepts/assets/software-defined-assets#viewing-assets-in-dagit).

Assets help track and define cross-job dependencies. For example, when viewing a job that materializes assets, you can navigate to the jobs that produce the assets that it depends on. Additionally, when an upstream asset has been updated more recently than a downstream asset, Dagster will indicate that the downstream asset might be out of date.

### Direct operation

Using software-defined assets enables you to directly operate them in Dagit. On the [Assets page](/concepts/dagit/dagit#assets), you can:

- View the materialization history of an asset
- Check when the next materialization will occur
- Launch a run to materialize the asset

Drilling down to the [Asset Details page](/concepts/dagit/dagit#asset-details) for an individual asset shows the sensors or schedules for any jobs targeting the asset. It also includes a **(Re)Materialize** button, allowing you to launch a run that rematerializes the asset, including its ancestors or descendants.

> ### Improved code ergonomics

Author: The way I was envisioning the structure of this section was that there are two separate top-level reasons that SDAs have "Improved Code Ergonomics". One is that using software-defined assets often means writing less code. The other is that you get to avoid a tradeoff between easy dependency tracking and manageable organization.

Not sure if my version did a good job, but I think the current structure mixes these in a kind of confusing way.

Reviewer: Gotcha! I admit I didn't catch that in the original version, but I see it now that you've clarified. I'll work some magic.

Without software-defined assets, you're often forced to choose between:

- **Easy dependency tracking** by containing everything in a single mega-job, which requires maintaining a huge, unwieldy object.

- **Manageable maintenance** by splitting up your pipeline into smaller jobs, which makes tracking cross-job dependencies difficult.

This isn't the case with software-defined assets:

- **Assets track their dependencies,** so you don't need to define mega-jobs just to keep track. This allows you to avoid interruptions in dependency graphs and eliminate the need for root input managers.

- **Using software-defined assets usually means writing less code**. Specifying the inputs to a software-defined asset defines the assets it depends on. This means you don't need to use `@graph` and `@job` to wire dependencies between ops.

>   This approach reduces the number of times an asset's name appears in your codebase by half, making improving your codebase's scalability. Refer to the [IO manager-based example](#materialize-two-interdependent-tables-without-an-io-manager) below to see this in action.

Reviewer 2: ```suggestion
  This approach reduces the number of times an asset's name appears in your codebase by half, improving your codebase's scalability. Refer to the [IO manager-based example](#materialize-two-interdependent-tables-without-an-io-manager) below to see this in action.
```

---

## When is it appropriate to use software-defined assets?

You should use software-defined assets when:

- You’re using Dagster to produce or maintain assets, AND

> - You know what those assets will be before you launch any runs. Note that using software-defined assets in one job doesn’t mean they must be used in all your jobs.

Author: "Note that using software-defined assets in one job doesn’t mean they must be used in all your jobs." is a separate point that I don't think is related to this bullet.

If your use case doesn't meet these criteria, you can still use graphs and ops.

Still not sure? Check out these examples to see what's a good fit and what isn't:

> | Use case                                                                                                 | Good fit? | Explanation                                                      |

Author: I like this approach of putting everything together in the table.

| Every day, drop and recreate the `users` table and the `user_recommender_model` model that depends on it | Yes       | Assets are known before a run and are being updated              |
| Every hour, add a partition to the `events` table                                                        | Yes       | Assets are known before a run and are being updated              |
| Clicking a button refreshes the `recommender` model                                                      | Yes       | Assets are known before a run and are being updated              |
| Every day, send emails to a set of users                                                                 | No        | No assets are being updated                                      |
| Every day, read a file of user IDs and change the value of a particular attribute for each user          | No        | The set of assets to update is not known before running the job. |
| Every day, scan my warehouse for tables that haven't been used in months and delete them                 | No        | The set of assets to update is not known before running the job. |

---

> ## How do software-defined assets work with other Dagster concepts?

Reviewer: @sryza This is what I wound up doing with what was previously the `Concept mapping` section. I also took the liberty of grouping concepts together to make the info progress the same way it does in the sidenav (for now) and be a bit easier to parse.

Author: I think breaking it up into these sections is really helpful. I still think that this should go below the migration examples. IMO the migration examples are the heart of this document and what I would be trying to get to as fast as possible if I was reading this.  Reading this stuff first would stress me out that there's all this stuff I need to think about to understand software-defined assets.

Reviewer: Agreed!

I think I'm getting hung up on how including this as the last section will break the 'flow' the current structure has. Specifically, the first few headings are questions (ex: `Why use assets?`) and the actual example section deviates from it (`Upgrading jobs...`)

I know your original draft sorta went this direction (1How to upgrade to...`) but it wasn't phrased as a question. What if we went with `How do I upgrade my jobs?` instead of `Upgrading jobs to...`? 

That way, we can move the example content up and have it be more of the star, the way it should be. Thoughts?

Author: > I think I'm getting hung up on how including this as the last section will break the 'flow' the current structure has. Specifically, the first few headings are questions (ex: Why use assets?) and the actual example section deviates from it (Upgrading jobs...)

I hear you

> I know your original draft sorta went this direction (How to upgrade to...) but it wasn't phrased as a question. What if we went with How do I upgrade my jobs? instead of Upgrading jobs to...?

Sounds great - I think that's better than the heading in my original draft.

Reviewer: Done and done.

| ------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------- |
| An [op]() is the basic unit of computation                                | Every software-defined asset includes a graph or an op                                         |
| A [graph]() is a composite unit of computation that connects multiple ops | Every software-defined asset includes a graph or an op                                         |
| Ops can have multiple outputs                                             | Multiple assets can be produced by a single op when defined using the `@multi_asset` decorator |
| Ops can use [config](/concepts/ops-jobs-graphs/ops#op-configuration)      | Assets can use [config]()                                                                      |
| Ops can be tested by directly invoking them                               | Assets can be tested by directly invoking them                                                 |

> | Ops can require resources                                                 | Software-defined assets can require resources                                                  |

Reviewer 2: I think it would be helpful here to mention that ops & software defined assets both have access to `OpExecutionContext`

from pandas import DataFrame

from dagster import SourceAsset, asset, define_asset_job, repository, with_resources

from .mylib import s3_io_manager, snowflake_io_manager, train_recommender_model

raw_users = SourceAsset(key="raw_users", io_manager_key="warehouse")

@asset(io_manager_key="warehouse")
def users(raw_users: DataFrame) -> DataFrame:
    users_df = raw_users.dropna()
    return users_df

@asset(io_manager_key="object_store")
def user_recommender_model(users: DataFrame):
    users_recommender_model = train_recommender_model(users)
    return users_recommender_model

@repository
def repo():
    return [
        *with_resources(
            [raw_users, users, user_recommender_model],
            resource_defs={
                "warehouse": snowflake_io_manager,
                "object_store": s3_io_manager,
            },
        ),
        define_asset_job("users_recommender_job"),
    ]
```

### Materialize two interdependent tables without an IO manager

- [Without software-defined-assets](#without-software-defined-assets-1)
- [With software-defined assets](#with-software-defined-assets-1)

#### Without software-defined assets

> This example does the same things as the [previous example](#a-job-that-materializes-two-interdependent-tables), with one difference. This job performs IO inside of the ops instead of delegating it to IO managers and root input managers:

Reviewer: ```suggestion
This example does the same things as the [previous example](#materialize-two-interdependent-tables), with one difference. This job performs IO inside of the ops instead of delegating it to IO managers and root input managers:
```
